SSA/Projects/FullyHomomorphicEncryption/Syntax.lean

```diff
@@ -0,0 +1,274 @@
+import SSA.Projects.InstCombine.LLVM.Transform
+import SSA.Projects.MLIRSyntax.AST
+import SSA.Projects.MLIRSyntax.EDSL
+import SSA.Projects.FullyHomomorphicEncryption.Basic
+
+open MLIR AST Ctxt
+open Polynomial -- for R[X] notation
+
+def ROfZComputable_impl (z : ℤ) : R q n :=
+  let zq : ZMod q := z
+  let p : (ZMod q)[X] := {
+      toFinsupp := Finsupp.mk
+        (support := List.toFinset (if zq = 0 then  [] else [0]))
+        (toFun := fun i => if i = 0 then z else 0)
+        (mem_support_toFun := by
+          intros a
+          constructor
+          · intros ha
+            simp at ha
+            split at ha
+            case mp.inl hz =>
+              simp at ha
+            case mp.inr hz =>
+              simp at ha
+              subst ha
+              simp [hz]
+          · intros hb
+            simp at hb
+            obtain ⟨ha, hz⟩ := hb
+            subst ha
+            simp [hz]
+        )
+      : (ZMod q)[X]
+  }
+  R.fromPoly p
+
+theorem ROfZComputable_eq (q n : Nat) (z : ℤ):
+    ROfZComputable_impl z = (↑z : R q n) := by
+  simp (config := { decide := true}) [ROfZComputable_impl] -- , Int.cast, Ring.toIntCast, IntCast.intCast]
+  by_cases (z = 0)
+  case pos h =>
+    simp (config := { decide := true, eta := true}) [h, EmptyCollection.emptyCollection]
+    have heta : forall a: Nat, ((fun (i : Nat) =>
+                      @OfNat.ofNat.{0} (ZMod q) 0
+                        (@Zero.toOfNat0.{0} (ZMod q)
+                          (@NegZeroClass.toZero.{0} (ZMod q)
+                            (@SubNegZeroMonoid.toNegZeroClass.{0} (ZMod q)
+                              (@SubtractionMonoid.toSubNegZeroMonoid.{0} (ZMod q)
+                                (@AddGroup.toSubtractionMonoid.{0} (ZMod q)
+                                  (@AddGroupWithOne.toAddGroup.{0} (ZMod q)
+                                    (@Ring.toAddGroupWithOne.{0} (ZMod q)
+                                      (@CommRing.toRing.{0} (ZMod q) (ZMod.commRing q))))))))))
+                    a)  = (0 : ZMod q) := by simp
+    set_option pp.all true in
+    try rw [heta] -- literally copy-pasted!
+    sorry
```

Comment:
I am a tad worry about the `sorry`.

Reply:
that's your hack to get around the noncomputable stuff leaking into the elaborator 😅